### PR TITLE
Refactor #5580: Add new file&directory apis into FileUtils and test cases

### PR DESCRIPTION
**Pull request by pandamicro**

### CODE REVIEW

File: cocos/platform/CCFileUtils.h [hunk 1]
```diff
@@ -284,6 +284,30 @@ class CC_DLL FileUtils
...
+     *  Write a ValueMap to a plist file.
+     *  @note This method is used internally.
+     */
+    virtual bool writeToFile(ValueMap& dict, const std::string& fullPath);
+    
+    /**
+     *  Converts the contents of a file to a ValueVector.
+     *  @note This method is used internally.
+     */
+    virtual ValueVector getValueVectorFromFile(const std::string& filename);
```
Comment (minggo) [inline cocos/platform/CCFileUtils.h:308]: Please input internally functions into `private` or `protected`.

**Comment by minggo** [comment]
I think new functions should not only support absolute path. It should support the path relative to the root of an application.

**Comment by pandamicro** [comment]
@minggo I think isDirectoryExist should support relative path and this have been done. But not other manipulation functions, because user need to know exactly which directory / file to remove / add, a relative path didn't identify exactly one directory or file.

File: cocos/platform/CCFileUtils.cpp [hunk 2]
```diff
@@ -916,6 +936,255 @@ bool FileUtils::isAbsolutePath(const std::string& path) const
...
+    }
+    return true;
+#else
+    if ((GetFileAttributesA(path.c_str())) == INVALID_FILE_ATTRIBUTES)
+    {
+		subpath = "";
+		for(int i = 0 ; i < dirs.size() ; ++i)
+		{
+			subpath += dirs[i];
+			BOOL ret = CreateDirectoryA(subpath.c_str(), NULL);
```
Comment (minggo) [inline cocos/platform/CCFileUtils.cpp:1052]: Some issue:
- BOOL -> bool
- NULL -> nullptr
- why not implement this function based on `FileUtils::createDirectory`?

Comment (minggo) [inline cocos/platform/CCFileUtils.cpp:1052]: Another one, please replace tab with 4 spaces.

File: tests/cpp-tests/Classes/FileUtilsTest/FileUtilsTest.cpp [hunk 3]
```diff
@@ -307,6 +309,168 @@ std::string TestIsFileExist::subtitle() const
...
+    std::string msg;
+    
+    FILE *out = fopen(filepath.c_str(), "w");
+    fputs(content.c_str(), out);
+    fclose(out);
+    
+    // Check whether file can be created
+    if (sharedFileUtils->isFileExist(filepath))
+    {
+        pTTF = Label::createWithSystemFont("Test file '__test.test' created", "", 20);
```
Comment (minggo) [inline tests/cpp-tests/Classes/FileUtilsTest/FileUtilsTest.cpp]: pTTF -> TTF.
Don't use Hungary style.

**Comment by pandamicro** [comment]
@minggo Thanks for the suggestions, it's done now, please refer to the last commit

**Comment by minggo** [comment]
Thanks.
Could we merge `createDirectory` and `createDirectories`?
It is not convenient to has two functions.

**Comment by pandamicro** [comment]
@minggo I agree with you, and it has been done in the new commit

File: cocos/platform/CCFileUtils.h [hunk 4]
```diff
@@ -384,6 +439,9 @@ class CC_DLL FileUtils
     virtual std::string getFullPathForDirectoryAndFilename(const std::string& directory, const std::string& filename);
     
     
+    virtual std::string searchFullPathForFilename(const std::string& filename) const;
```
Comment (minggo) [inline cocos/platform/CCFileUtils.h:442]: What's this function mean?

File: cocos/platform/CCFileUtils.h [hunk 5]
```diff
@@ -362,6 +412,11 @@ class CC_DLL FileUtils
     virtual bool isFileExistInternal(const std::string& filename) const = 0;
     
     /**
+     *  Checks whether file exists without considering search paths and resolution orders.
+     */
+    virtual bool isDirectoryExistInternal(const std::string& dirPath) const;
```
Comment (minggo) [inline cocos/platform/CCFileUtils.h:417]: I don't think we need this function.

File: cocos/platform/CCFileUtils.h (hunk 4 above)
Comment (pandamicro) [inline cocos/platform/CCFileUtils.h:442]: To detect the full path of a given filename regarding all the search path and resolution

File: cocos/platform/CCFileUtils.h (hunk 5 above)
Comment (pandamicro) [inline cocos/platform/CCFileUtils.h:417]: This is for keeping the `isDirectoryExist` function clean

**Pull request merged**